{
  "title": "howto-gdb_helpers",
  "language": "python",
  "topics": [
    "machine_learning",
    "web_dev",
    "fundamentals",
    "algorithms",
    "data_structures",
    "networking",
    "database"
  ],
  "purpose": "diagnose and correct the issue.",
  "code": "stack of currently executing Python functions. Given a Python object represented by a PyObject* pointer, the\nextensionsurfacesthetypeandvalueoftheobject.\nDeveloperswhoareworkingonCPythonextensionsortinkeringwithpartsofCPythonthatarewritteninCcanuse\nthisdocumenttolearnhowtousethepython-gdb.pyextensionwithGDB.\n(cid:174) Note\nThis document assumes that you are familiar with the basics of GDB and the CPython C API. It consolidates\nguidancefromthedevguideandthePythonwiki.\n1\n1 Prerequisites\nYouneedtohave:\n\u2022 GDB7orlater. (ForearlierversionsofGDB,seeMisc/gdbinitinthesourcesofPython3.11orearlier.)\n\u2022 GDB-compatibledebugginginformationforPythonandanyextensionyouaredebugging.\n\u2022 Thepython-gdb.pyextension.\nTheextensionisbuiltwithPython,butmightbedistributedseparatelyornotatall. Below,weincludetipsforafew\ncommonsystemsasexamples. Notethateveniftheinstructionsmatchyoursystem,theymightbeoutdated.\n1.1 Setup with Python built from source\nWhen you build CPython from source, debugging information should be available, and the build should add a\npython-gdb.pyfiletotherootdirectoryofyourrepository.\nToactivatesupport,youmustaddthedirectorycontainingpython-gdb.pytoGDB\u2019s\u201cauto-load-safe-path\u201d. Ifyou\nhaven\u2019tdonethis,recentversionsofGDBwillprintoutawarningwithinstructionsonhowtodothis.\n(cid:174) Note\nIfyoudonotseeinstructionsforyourversionofGDB,putthisinyourconfigurationfile(~/.gdbinitor~/.\nconfig/gdb/gdbinit):\nadd-auto-load-safe-path /path/to/cpython\nYoucanalsoaddmultiplepaths,separatedby:.\n1.2 Setup for Python from a Linux distro\nMost Linux systems provide debug information for the system Python in a package called python-debuginfo,\npython-dbgorsimilar. Forexample:\n\u2022 Fedora:\nsudo dnf install gdb\nsudo dnf debuginfo-install python3\n\u2022 Ubuntu:\nsudo apt install gdb python3-dbg\nOnseveralrecentLinuxsystems,GDBcandownloaddebuggingsymbolsautomaticallyusingdebuginfod. However,\nthiswillnotinstallthepython-gdb.pyextension;yougenerallydoneedtoinstallthedebuginfopackageseparately.\n2 Using the Debug build and Development mode\nForeasierdebugging,youmightwantto:\n\u2022 UseadebugbuildofPython. (Whenbuildingfromsource,useconfigure --with-pydebug. OnLinux\ndistros,installandrunapackagelikepython-debugorpython-dbg,ifavailable.)\n\u2022 Usetheruntimedevelopmentmode(-X dev).\nBothenableextraassertionsanddisablesomeoptimizations. Sometimesthishidesthebugyouaretryingtofind,but\ninmostcasestheymaketheprocesseasier.\n2\n3 Using the python-gdb extension\nWhentheextensionisloaded,itprovidestwomainfeatures: prettyprintersforPythonvalues,andadditionalcom-\nmands.\n3.1 Pretty-printers\nThisiswhataGDBbacktracelookslike(truncated)whenthisextensionisenabled:\n#0 0x000000000041a6b1 in PyObject_Malloc (nbytes=Cannot access memory at address\u2423\n,\u21920x7fffff7fefe8\n) at Objects/obmalloc.c:748\n#1 0x000000000041b7c0 in _PyObject_DebugMallocApi (id=111 'o', nbytes=24) at\u2423\n,\u2192Objects/obmalloc.c:1445\n#2 0x000000000041b717 in _PyObject_DebugMalloc (nbytes=24) at Objects/obmalloc.\n,\u2192c:1412\n#3 0x000000000044060a in _PyUnicode_New (length=11) at Objects/unicodeobject.c:346\n#4 0x00000000004466aa in PyUnicodeUCS2_DecodeUTF8Stateful (s=0x5c2b8d \"__lltrace__\n,\u2192\", size=11, errors=0x0, consumed=\n0x0) at Objects/unicodeobject.c:2531\n#5 0x0000000000446647 in PyUnicodeUCS2_DecodeUTF8 (s=0x5c2b8d \"__lltrace__\",\u2423\n,\u2192size=11, errors=0x0)\nat Objects/unicodeobject.c:2495\n#6 0x0000000000440d1b in PyUnicodeUCS2_FromStringAndSize (u=0x5c2b8d \"__lltrace__\n,\u2192\", size=11)\nat Objects/unicodeobject.c:551\n#7 0x0000000000440d94 in PyUnicodeUCS2_FromString (u=0x5c2b8d \"__lltrace__\") at\u2423\n,\u2192Objects/unicodeobject.c:569\n#8 0x0000000000584abd in PyDict_GetItemString (v=\n{'Yuck': <type at remote 0xad4730>, '__builtins__': <module at remote\u2423\n,\u21920x7ffff7fd5ee8>, '__file__': 'Lib/test/crashers/nasty_eq_vs_dict.py', '__package_\n,\u2192_': None, 'y': <Yuck(i=0) at remote 0xaacd80>, 'dict': {0: 0, 1: 1, 2: 2, 3: 3},\n,\u2192'__cached__': None, '__name__': '__main__', 'z': <Yuck(i=0) at remote 0xaace60>,\n,\u2192'__doc__': None}, key=\n0x5c2b8d \"__lltrace__\") at Objects/dictobject.c:2171\nNoticehowthedictionaryargumenttoPyDict_GetItemStringisdisplayedasitsrepr(),ratherthananopaque\nPyObject *pointer.\nTheextensionworksbysupplyingacustomprintingroutineforvaluesoftypePyObject *. Ifyouneedtoaccess\nlower-leveldetailsofanobject,thencastthevaluetoapointeroftheappropriatetype. Forexample:\n(gdb) p globals\n$1 = {'__builtins__': <module at remote 0x7ffff7fb1868>, '__name__':\n'__main__', 'ctypes': <module at remote 0x7ffff7f14360>, '__doc__': None,\n'__package__': None}\n(gdb) p *(PyDictObject*)globals\n$2 = {ob_refcnt = 3, ob_type = 0x3dbdf85820, ma_fill = 5, ma_used = 5,\nma_mask = 7, ma_table = 0x63d0f8, ma_lookup = 0x3dbdc7ea70\n<lookdict_string>, ma_smalltable = {{me_hash = 7065186196740147912,\nme_key = '__builtins__', me_value = <module at remote 0x7ffff7fb1868>},\n{me_hash = -368181376027291943, me_key = '__name__',\nme_value ='__main__'}, {me_hash = 0, me_key = 0x0, me_value = 0x0},\n{me_hash = 0, me_key = 0x0, me_value = 0x0},\n{me_hash = -9177857982131165996, me_key = 'ctypes',\nme_value = <module at remote 0x7ffff7f14360>},\n{me_hash = -8518757509529533123, me_key = '__doc__', me_value = None},\n(continuesonnextpage)\n3\n(continuedfrompreviouspage)\n{me_hash = 0, me_key = 0x0, me_value = 0x0}, {\nme_hash = 6614918939584953775, me_key = '__package__', me_value = None}}}\nNotethatthepretty-printersdonotactuallycallrepr(). Forbasictypes,theytrytomatchitsresultclosely.\nAn area that can be confusing is that the custom printer for some types look a lot like GDB\u2019s built-in printer for\nstandardtypes. Forexample,thepretty-printerforaPythonint(PyLongObject*)givesarepresentationthatis\nnotdistinguishablefromoneofaregularmachine-levelinteger:\n(gdb) p some_machine_integer\n$3 = 42\n(gdb) p some_python_integer\n$4 = 42\nTheinternalstructurecanberevealedwithacasttoPyLongObject*:\n(gdb) p *(PyLongObject*)some_python_integer\n$5 = {ob_base = {ob_base = {ob_refcnt = 8, ob_type = 0x3dad39f5e0}, ob_size = 1},\nob_digit = {42}}\nAsimilarconfusioncanarisewiththestrtype,wheretheoutputlooksalotlikegdb\u2019sbuilt-inprinterforchar *:\n(gdb) p ptr_to_python_str\n$6 = '__builtins__'\nThepretty-printerforstrinstancesdefaultstousingsingle-quotes(asdoesPython\u2019sreprforstrings)whereasthe\nstandardprinterforchar *valuesusesdouble-quotesandcontainsahexadecimaladdress:\n(gdb) p ptr_to_char_star\n$7 = 0x6d72c0 \"hello world\"\nAgain,theimplementationdetailscanberevealedwithacasttoPyUnicodeObject*:\n(gdb) p *(PyUnicodeObject*)$6\n$8 = {ob_base = {ob_refcnt = 33, ob_type = 0x3dad3a95a0}, length = 12,\nstr = 0x7ffff2128500, hash = 7065186196740147912, state = 1, defenc = 0x0}\n3.2 py-list\nTheextensionaddsapy-listcommand, whichliststhePythonsourcecode(ifany)forthecurrent\nframeintheselectedthread. Thecurrentlineismarkedwitha\u201c>\u201d:\n(gdb) py-list\n901 if options.profile:\n902 options.profile = False\n903 profile_me()\n904 return\n905\n>906 u = UI()\n907 if not u.quit:\n908 try:\n909 gtk.main()\n910 except KeyboardInterrupt:\n911 # properly quit on a keyboard interrupt...\nUsepy-list STARTtolistatadifferentlinenumberwithinthePythonsource,andpy-list START,\nENDtolistaspecificrangeoflineswithinthePythonsource.\n4\n3.3 py-up and py-down\nThepy-upandpy-downcommandsareanalogoustoGDB\u2019sregularupanddowncommands,buttry\ntomoveatthelevelofCPythonframes,ratherthanCframes.\nGDBisnotalwaysabletoreadtherelevantframeinformation,dependingontheoptimizationlevelwith\nwhichCPythonwascompiled. Internally,thecommandslookforCframesthatareexecutingthedefault\nframeevaluationfunction(thatis,thecorebytecodeinterpreterloopwithinCPython)andlookupthe\nvalueoftherelatedPyFrameObject *.\nTheyemittheframenumber(attheClevel)withinthethread.\nForexample:\n(gdb) py-up\n#37 Frame 0x9420b04, for file /usr/lib/python2.6/site-packages/\ngnome_sudoku/main.py, line 906, in start_game ()\nu = UI()\n(gdb) py-up\n#40 Frame 0x948e82c, for file /usr/lib/python2.6/site-packages/\ngnome_sudoku/gnome_sudoku.py, line 22, in start_game(main=<module at\u2423\n,\u2192remote 0xb771b7f4>)\nmain.start_game()\n(gdb) py-up\nUnable to find an older python frame\nsowe\u2019reatthetopofthePythonstack.\nThe frame numbers correspond to those displayed by GDB\u2019s standard backtrace command. The\ncommandskipsCframeswhicharenotexecutingPythoncode.\nGoingbackdown:\n(gdb) py-down\n#37 Frame 0x9420b04, for file /usr/lib/python2.6/site-packages/gnome_\n,\u2192sudoku/main.py, line 906, in start_game ()\nu = UI()\n(gdb) py-down\n#34 (unable to read python frame information)\n(gdb) py-down\n#23 (unable to read python frame information)\n(gdb) py-down\n#19 (unable to read python frame information)\n(gdb) py-down\n#14 Frame 0x99262ac, for file /usr/lib/python2.6/site-packages/gnome_\n,\u2192sudoku/game_selector.py, line 201, in run_swallowed_dialog (self=\n,\u2192<NewOrSavedGameSelector(new_game_model=<gtk.ListStore at remote\u2423\n,\u21920x98fab44>, puzzle=None, saved_games=[{'gsd.auto_fills': 0, 'tracking':\n,\u2192{}, 'trackers': {}, 'notes': [], 'saved_at': 1270084485, 'game': '7 8 0\u2423\n,\u21920 0 0 0 5 6 0 0 9 0 8 0 1 0 0 0 4 6 0 0 0 0 7 0 6 5 0 0 0 4 7 9 2 0 0 0\u2423\n,\u21929 0 1 0 0 0 3 9 7 6 0 0 0 1 8 0 6 0 0 0 0 2 8 0 0 0 5 0 4 0 6 0 0 2 1 0\u2423\n,\u21920 0 0 0 4 5\\n7 8 0 0 0 0 0 5 6 0 0 9 0 8 0 1 0 0 0 4 6 0 0 0 0 7 0 6 5\u2423\n,\u21921 8 3 4 7 9 2 0 0 0 9 0 1 0 0 0 3 9 7 6 0 0 0 1 8 0 6 0 0 0 0 2 8 0 0 0\u2423\n,\u21925 0 4 0 6 0 0 2 1 0 0 0 0 0 4 5', 'gsd.impossible_hints': 0, 'timer.__\n,\u2192absolute_start_time__': <float at remote 0x984b474>, 'gsd.hints': 0,\n,\u2192'timer.active_time': <float at remote 0x984b494>, 'timer.total_time':\n,\u2192<float at remote 0x984b464>}], dialog=<gtk.Dialog at remote 0x98faaa4>,\u2423\n,\u2192saved_game_model=<gtk.ListStore at remote 0x98fad24>, sudoku_maker=\n,\u2192<SudokuMaker(terminated=False, played=[], batch_siz...(truncated)\nswallower.run_dialog(self.dialog)\n(gdb) py-down\n(continuesonnextpage)\n5\n(continuedfrompreviouspage)\n#11 Frame 0x9aead74, for file /usr/lib/python2.6/site-packages/gnome_\n,\u2192sudoku/dialog_swallower.py, line 48, in run_dialog (self=\n,\u2192<SwappableArea(running=<gtk.Dialog at remote 0x98faaa4>, main_page=0)\u2423\n,\u2192at remote 0x98fa6e4>, d=<gtk.Dialog at remote 0x98faaa4>)\ngtk.main()\n(gdb) py-down\n#8 (unable to read python frame information)\n(gdb) py-down\nUnable to find a newer python frame\nandwe\u2019reatthebottomofthePythonstack.\nNote that in Python 3.12 and newer, the same C stack frame can be used for multiple Python stack\nframes. Thismeansthatpy-upandpy-downmaymovemultiplePythonframesatonce. Forexample:\n(gdb) py-up\n#6 Frame 0x7ffff7fb62b0, for file /tmp/rec.py, line 5, in recursive_\n,\u2192function (n=0)\ntime.sleep(5)\n#6 Frame 0x7ffff7fb6240, for file /tmp/rec.py, line 7, in recursive_\n,\u2192function (n=1)\nrecursive_function(n-1)\n#6 Frame 0x7ffff7fb61d0, for file /tmp/rec.py, line 7, in recursive_\n,\u2192function (n=2)\nrecursive_function(n-1)\n#6 Frame 0x7ffff7fb6160, for file /tmp/rec.py, line 7, in recursive_\n,\u2192function (n=3)\nrecursive_function(n-1)\n#6 Frame 0x7ffff7fb60f0, for file /tmp/rec.py, line 7, in recursive_\n,\u2192function (n=4)\nrecursive_function(n-1)\n#6 Frame 0x7ffff7fb6080, for file /tmp/rec.py, line 7, in recursive_\n,\u2192function (n=5)\nrecursive_function(n-1)\n#6 Frame 0x7ffff7fb6020, for file /tmp/rec.py, line 9, in <module> ()\nrecursive_function(5)\n(gdb) py-up\nUnable to find an older python frame\n3.4 py-bt\nThepy-btcommandattemptstodisplayaPython-levelbacktraceofthecurrentthread.\nForexample:\n(gdb) py-bt\n#8 (unable to read python frame information)\n#11 Frame 0x9aead74, for file /usr/lib/python2.6/site-packages/gnome_\n,\u2192sudoku/dialog_swallower.py, line 48, in run_dialog (self=\n,\u2192<SwappableArea(running=<gtk.Dialog at remote 0x98faaa4>, main_page=0)\u2423\n,\u2192at remote 0x98fa6e4>, d=<gtk.Dialog at remote 0x98faaa4>)\ngtk.main()\n#14 Frame 0x99262ac, for file /usr/lib/python2.6/site-packages/gnome_\n,\u2192sudoku/game_selector.py, line 201, in run_swallowed_dialog (self=\n,\u2192<NewOrSavedGameSelector(new_game_model=<gtk.ListStore at remote\u2423\n,\u21920x98fab44>, puzzle=None, saved_games=[{'gsd.auto_fills': 0, 'tracking':\n,\u2192{}, 'trackers': {}, 'notes': [], 'saved_at': 1270084485, 'game': '7 8 0\u2423\n(continuesonnextpage)\n6\n(continuedfrompreviouspage)\n,\u21920 0 0 0 5 6 0 0 9 0 8 0 1 0 0 0 4 6 0 0 0 0 7 0 6 5 0 0 0 4 7 9 2 0 0 0\u2423\n,\u21929 0 1 0 0 0 3 9 7 6 0 0 0 1 8 0 6 0 0 0 0 2 8 0 0 0 5 0 4 0 6 0 0 2 1 0\u2423\n,\u21920 0 0 0 4 5\\n7 8 0 0 0 0 0 5 6 0 0 9 0 8 0 1 0 0 0 4 6 0 0 0 0 7 0 6 5\u2423\n,\u21921 8 3 4 7 9 2 0 0 0 9 0 1 0 0 0 3 9 7 6 0 0 0 1 8 0 6 0 0 0 0 2 8 0 0 0\u2423\n,\u21925 0 4 0 6 0 0 2 1 0 0 0 0 0 4 5', 'gsd.impossible_hints': 0, 'timer.__\n,\u2192absolute_start_time__': <float at remote 0x984b474>, 'gsd.hints': 0,\n,\u2192'timer.active_time': <float at remote 0x984b494>, 'timer.total_time':\n,\u2192<float at remote 0x984b464>}], dialog=<gtk.Dialog at remote 0x98faaa4>,\u2423\n,\u2192saved_game_model=<gtk.ListStore at remote 0x98fad24>, sudoku_maker=\n,\u2192<SudokuMaker(terminated=False, played=[], batch_siz...(truncated)\nswallower.run_dialog(self.dialog)\n#19 (unable to read python frame information)\n#23 (unable to read python frame information)\n#34 (unable to read python frame information)\n#37 Frame 0x9420b04, for file /usr/lib/python2.6/site-packages/gnome_\n,\u2192sudoku/main.py, line 906, in start_game ()\nu = UI()\n#40 Frame 0x948e82c, for file /usr/lib/python2.6/site-packages/gnome_\n,\u2192sudoku/gnome_sudoku.py, line 22, in start_game (main=<module at remote\u2423\n,\u21920xb771b7f4>)\nmain.start_game()\nTheframenumberscorrespondtothosedisplayedbyGDB\u2019sstandardbacktracecommand.\n3.5 py-print\nThe py-print command looks up a Python name and tries to print it. It looks in locals within the\ncurrentthread,thenglobals,thenfinallybuiltins:\n(gdb) py-print self\nlocal 'self' = <SwappableArea(running=<gtk.Dialog at remote 0x98faaa4>,\nmain_page=0) at remote 0x98fa6e4>\n(gdb) py-print __name__\nglobal '__name__' = 'gnome_sudoku.dialog_swallower'\n(gdb) py-print len\nbuiltin 'len' = <built-in function len>\n(gdb) py-print scarlet_pimpernel\n'scarlet_pimpernel' not found\nIfthecurrentCframecorrespondstomultiplePythonframes,py-printonlyconsidersthefirstone.\n3.6 py-locals\nThepy-localscommandlooksupallPythonlocalswithinthecurrentPythonframeintheselected\nthread,andprintstheirrepresentations:\n(gdb) py-locals\nself = <SwappableArea(running=<gtk.Dialog at remote 0x98faaa4>,\nmain_page=0) at remote 0x98fa6e4>\nd = <gtk.Dialog at remote 0x98faaa4>\nIfthecurrentCframecorrespondstomultiplePythonframes,localsfromallofthemwillbeshown:\n(gdb) py-locals\nLocals for recursive_function\nn = 0\nLocals for recursive_function\n(continuesonnextpage)\n7\n(continuedfrompreviouspage)\nn = 1\nLocals for recursive_function\nn = 2\nLocals for recursive_function\nn = 3\nLocals for recursive_function\nn = 4\nLocals for recursive_function\nn = 5\nLocals for <module>\n4 Use with GDB commands\nTheextensioncommandscomplementGDB\u2019sbuilt-incommands. Forexample,youcanuseaframenumbersshown\nbypy-btwiththeframecommandtogoaspecificframewithintheselectedthread,likethis:\n(gdb) py-bt\n(output snipped)\n#68 Frame 0xaa4560, for file Lib/test/regrtest.py, line 1548, in <module> ()\nmain()\n(gdb) frame 68\n#68 0x00000000004cd1e6 in PyEval_EvalFrameEx (f=Frame 0xaa4560, for file Lib/test/\n,\u2192regrtest.py, line 1548, in <module> (), throwflag=0) at Python/ceval.c:2665\n2665 x = call_function(&sp, oparg);\n(gdb) py-list\n1543 # Run the tests in a context manager that temporary changes the CWD to\u2423\n,\u2192a\n1544 # temporary and writable directory. If it's not possible to create or\n1545 # change the CWD, the original CWD will be used. The original CWD is\n1546 # available from test_support.SAVEDCWD.\n1547 with test_support.temp_cwd(TESTCWD, quiet=True):\n>1548 main()\nTheinfo threadscommandwillgiveyoualistofthethreadswithintheprocess,andyoucanusethethread\ncommandtoselectadifferentone:\n(gdb) info threads\n105 Thread 0x7fffefa18710 (LWP 10260) sem_wait () at ../nptl/sysdeps/unix/sysv/\n,\u2192linux/x86_64/sem_wait.S:86\n104 Thread 0x7fffdf5fe710 (LWP 10259) sem_wait () at ../nptl/sysdeps/unix/sysv/\n,\u2192linux/x86_64/sem_wait.S:86\n* 1 Thread 0x7ffff7fe2700 (LWP 10145) 0x00000038e46d73e3 in select () at ../\n,\u2192sysdeps/unix/syscall-template.S:82\nYoucanusethread apply all COMMANDor(t a a COMMANDforshort)torunacommandonallthreads. With\npy-bt,thisletsyouseewhateverythreadisdoingatthePythonlevel:\n(gdb) t a a py-bt\nThread 105 (Thread 0x7fffefa18710 (LWP 10260)):\n#5 Frame 0x7fffd00019d0, for file /home/david/coding/python-svn/Lib/threading.py,\u2423\n,\u2192line 155, in _acquire_restore (self=<_RLock(_Verbose__verbose=False, _RLock__\n,\u2192owner=140737354016512, _RLock__block=<thread.lock at remote 0x858770>, _RLock__\n,\u2192count=1) at remote 0xd7ff40>, count_owner=(1, 140737213728528), count=1,\u2423\n,\u2192owner=140737213728528)\nself.__block.acquire()\n(continuesonnextpage)\n8\n(continuedfrompreviouspage)\n#8 Frame 0x7fffac001640, for file /home/david/coding/python-svn/Lib/threading.py,\u2423\n,\u2192line 269, in wait (self=<_Condition(_Condition__lock=<_RLock(_Verbose__\n,\u2192verbose=False, _RLock__owner=140737354016512, _RLock__block=<thread.lock at\u2423\n,\u2192remote 0x858770>, _RLock__count=1) at remote 0xd7ff40>, acquire=<instancemethod\u2423\n,\u2192at remote 0xd80260>, _is_owned=<instancemethod at remote 0xd80160>, _release_\n,\u2192save=<instancemethod at remote 0xd803e0>, release=<instancemethod at remote\u2423\n,\u21920xd802e0>, _acquire_restore=<instancemethod at remote 0xd7ee60>, _Verbose__\n,\u2192verbose=False, _Condition__waiters=[]) at remote 0xd7fd10>, timeout=None, waiter=\n,\u2192<thread.lock at remote 0x858a90>, saved_state=(1, 140737213728528))\nself._acquire_restore(saved_state)\n#12 Frame 0x7fffb8001a10, for file /home/david/coding/python-svn/Lib/test/lock_\n,\u2192tests.py, line 348, in f ()\ncond.wait()\n#16 Frame 0x7fffb8001c40, for file /home/david/coding/python-svn/Lib/test/lock_\n,\u2192tests.py, line 37, in task (tid=140737213728528)\nf()\nThread 104 (Thread 0x7fffdf5fe710 (LWP 10259)):\n#5 Frame 0x7fffe4001580, for file /home/david/coding/python-svn/Lib/threading.py,\u2423\n,\u2192line 155, in _acquire_restore (self=<_RLock(_Verbose__verbose=False, _RLock__\n,\u2192owner=140737354016512, _RLock__block=<thread.lock at remote 0x858770>, _RLock__\n,\u2192count=1) at remote 0xd7ff40>, count_owner=(1, 140736940992272), count=1,\u2423\n,\u2192owner=140736940992272)\nself.__block.acquire()\n#8 Frame 0x7fffc8002090, for file /home/david/coding/python-svn/Lib/threading.py,\u2423\n,\u2192line 269, in wait (self=<_Condition(_Condition__lock=<_RLock(_Verbose__\n,\u2192verbose=False, _RLock__owner=140737354016512, _RLock__block=<thread.lock at\u2423\n,\u2192remote 0x858770>, _RLock__count=1) at remote 0xd7ff40>, acquire=<instancemethod\u2423\n,\u2192at remote 0xd80260>, _is_owned=<instancemethod at remote 0xd80160>, _release_\n,\u2192save=<instancemethod at remote 0xd803e0>, release=<instancemethod at remote\u2423\n,\u21920xd802e0>, _acquire_restore=<instancemethod at remote 0xd7ee60>, _Verbose__\n,\u2192verbose=False, _Condition__waiters=[]) at remote 0xd7fd10>, timeout=None, waiter=\n,\u2192<thread.lock at remote 0x858860>, saved_state=(1, 140736940992272))\nself._acquire_restore(saved_state)\n#12 Frame 0x7fffac001c90, for file /home/david/coding/python-svn/Lib/test/lock_\n,\u2192tests.py, line 348, in f ()\ncond.wait()\n#16 Frame 0x7fffac0011c0, for file /home/david/coding/python-svn/Lib/test/lock_\n,\u2192tests.py, line 37, in task (tid=140736940992272)\nf()\nThread 1 (Thread 0x7ffff7fe2700 (LWP 10145)):\n#5 Frame 0xcb5380, for file /home/david/coding/python-svn/Lib/test/lock_tests.py,\u2423\n,\u2192line 16, in _wait ()\ntime.sleep(0.01)\n#8 Frame 0x7fffd00024a0, for file /home/david/coding/python-svn/Lib/test/lock_\n,\u2192tests.py, line 378, in _check_notify (self=<ConditionTests(_testMethodName='test_\n,\u2192notify', _resultForDoCleanups=<TestResult(_original_stdout=<cStringIO.StringO at\u2423\n,\u2192remote 0xc191e0>, skipped=[], _mirrorOutput=False, testsRun=39, buffer=False, _\n,\u2192original_stderr=<file at remote 0x7ffff7fc6340>, _stdout_buffer=<cStringIO.\n,\u2192StringO at remote 0xc9c7f8>, _stderr_buffer=<cStringIO.StringO at remote\u2423\n,\u21920xc9c790>, _moduleSetUpFailed=False, expectedFailures=[], errors=[], _\n,\u2192previousTestClass=<type at remote 0x928310>, unexpectedSuccesses=[], failures=[],\n,\u2192 shouldStop=False, failfast=False) at remote 0xc185a0>, _threads=(0,), _\n,\u2192cleanups=[], _type_equality_funcs={<type at remote 0x7eba00>: <instancemethod at\u2423\n,\u2192remote 0xd750e0>, <type at remote 0x7e7820>: <instancemethod at remote 0xd75160>,\n(continuesonnextpage)\n9\n(continuedfrompreviouspage)\n,\u2192 <type at remote 0x7e30e0>: <instancemethod at remote 0xd75060>, <type at remote\u2423\n,\u21920x7e7d20>: <instancemethod at remote 0xd751e0>, <type at remote 0x7f19e0...\n,\u2192(truncated)\n_wait()\n10\n",
  "context": "diagnose and correct the issue. By default, GDB (or any of its front-ends) doesn\u2019t support high-level information\nspecifictotheCPythoninterpreter.\nThepython-gdb.pyextensionaddsCPythoninterpreterinformationtoGDB.Theextensionhelpsintrospectthe",
  "source_file": "resources\\Year 3\\Python\\howto-gdb_helpers.pdf",
  "line_numbers": [
    27,
    437
  ]
}